server/handlers/project_command_output_handler.go

```diff
@@ -83,32 +93,45 @@ func NewAsyncProjectCommandOutputHandler(
 		projectStatusUpdater:   projectStatusUpdater,
 		projectJobURLGenerator: projectJobURLGenerator,
 		projectOutputBuffers:   map[string][]string{},
+		pullToJobMapping:       map[models.PullContext]map[string]bool{},
 	}
 }
 
 func (p *AsyncProjectCommandOutputHandler) Send(ctx models.ProjectCommandContext, msg string) {
 	p.projectCmdOutput <- &models.ProjectCmdOutputLine{
-		ProjectInfo: ctx.PullInfo(),
-		Line:        msg,
+		JobID: ctx.JobID,
+		JobContext: models.JobContext{
+			HeadCommit: ctx.Pull.HeadCommit,
+			PullContext: models.PullContext{
+				PullNum:     ctx.Pull.Num,
+				Repo:        ctx.BaseRepo.Name,
+				ProjectName: ctx.ProjectName,
+				Workspace:   ctx.Workspace,
+			},
+		},
+		Line: msg,
 	}
 }
 
-func (p *AsyncProjectCommandOutputHandler) Register(projectInfo string, receiver chan string) {
-	p.addChan(receiver, projectInfo)
+func (p *AsyncProjectCommandOutputHandler) Register(jobID string, receiver chan string) {
+	p.addChan(receiver, jobID)
 }
 
 func (p *AsyncProjectCommandOutputHandler) Handle() {
 	for msg := range p.projectCmdOutput {
-		if msg.ClearBuffBefore {
-			p.clearLogLines(msg.ProjectInfo)
+		if !msg.ClearBuffBefore {
```

Comment:
There is no need to check this. You're generating new jobID for each command, so they're all should be already streaming to a clean buffer.

Reply:
Yup, thanks for the catch! I've removed this logic from the log handler as well as the templates!